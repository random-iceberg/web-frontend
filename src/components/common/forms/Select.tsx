import React from "react";

/**
 * Structure for options passed to the Select component.
 * @typedef {object} SelectOption
 * @property {string | number} value - The actual value submitted.
 * @property {string} label - The text displayed for the option.
 */
type SelectOption = {
  value: string | number;
  label: string;
};

// Omit the native 'onChange' from HTMLSelectAttributes to avoid type conflict
// with our custom 'onChange' that passes the string value directly.
/**
 * Props for the Select component. Extends standard HTML select attributes, omitting `onChange`.
 * @typedef {object} SelectProps
 * @property {string} label - Text label displayed above the select input.
 * @property {string} id - Unique ID for the select input and linking label.
 * @property {SelectOption[]} options - Array of options to populate the select dropdown.
 * @property {string | number} value - The currently selected value.
 * @property {(value: string) => void} onChange - Callback function when the selection changes. Passes the new string value.
 * @property {boolean} [required=false] - Whether the select input is required.
 * @property {string} [placeholder] - Optional placeholder text (rendered as a disabled option).
 * @property {boolean} [disabled=false] - Whether the select input is disabled.
 * @property {string} [className] - Additional CSS classes to apply.
 */
type SelectProps = Omit<
  React.SelectHTMLAttributes<HTMLSelectElement>,
  "onChange"
> & {
  label: string;
  id: string;
  options: SelectOption[];
  value: string | number;
  onChange: (value: string) => void; // Custom onChange handler
  required?: boolean;
  placeholder?: string; // Optional placeholder
};

/**
 * A styled wrapper around the native HTML `<select>` element.
 * Provides consistent styling with other form components and includes a label.
 *
 * @component
 * @param {SelectProps} props - Props for the Select component.
 * @example
 * const [algorithm, setAlgorithm] = useState('SVM');
 * const algoOptions = [
 *   { value: 'Random Forest', label: 'Random Forest' },
 *   { value: 'SVM', label: 'SVM' },
 * ];
 *
 * <Select
 *   id="algorithm-select"
 *   label="Algorithm"
 *   options={algoOptions}
 *   value={algorithm}
 *   onChange={setAlgorithm}
 *   required
 * />
 */
const Select: React.FC<SelectProps> = ({
  label,
  id,
  options,
  value,
  onChange,
  disabled = false,
  required = false,
  placeholder,
  className = "",
  ...props
}) => {
  return (
    <div className="w-full">
      <label
        htmlFor={id}
        className="block text-sm font-medium text-gray-700 mb-2"
      >
        {label}
        {required && <span className="text-red-500 ml-1">*</span>}
      </label>
      <select
        id={id}
        value={value}
        onChange={(e) => onChange(e.target.value)}
        disabled={disabled}
        required={required}
        className={`w-full rounded border-gray-300 shadow-sm text-sm px-4 py-2.5
          focus:ring-1 focus:ring-blue-500 focus:border-blue-500
          disabled:bg-gray-50 disabled:text-gray-500 disabled:cursor-not-allowed
          border ${className}`.trim()}
        {...props}
      >
        {placeholder && (
          <option value="" disabled>
            {placeholder}
          </option>
        )}
        {options.map((option) => (
          <option key={option.value} value={option.value}>
            {option.label}
          </option>
        ))}
      </select>
    </div>
  );
};

export default Select;
