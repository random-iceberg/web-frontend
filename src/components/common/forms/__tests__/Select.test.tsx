import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import Select from "../Select";

describe("Select Component", () => {
  const mockOnChange = jest.fn();
  const sampleOptions = [
    { value: "opt1", label: "Option 1" },
    { value: "opt2", label: "Option 2" },
    { value: "opt3", label: "Option 3" },
  ];

  beforeEach(() => {
    mockOnChange.mockClear();
  });

  it("renders correctly with basic props", () => {
    const { asFragment } = render(
      <Select
        id="test-select"
        label="Test Select"
        options={sampleOptions}
        value="opt1"
        onChange={mockOnChange}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly with a different selected value", () => {
    const { asFragment } = render(
      <Select
        id="test-select"
        label="Test Select"
        options={sampleOptions}
        value="opt3"
        onChange={mockOnChange}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly when disabled", () => {
    const { asFragment } = render(
      <Select
        id="test-select"
        label="Test Select"
        options={sampleOptions}
        value="opt1"
        onChange={mockOnChange}
        disabled
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly when required", () => {
    const { asFragment } = render(
      <Select
        id="test-select"
        label="Test Select"
        options={sampleOptions}
        value="opt1"
        onChange={mockOnChange}
        required
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders with a placeholder", () => {
    const { asFragment } = render(
      <Select
        id="test-select"
        label="Test Select"
        options={sampleOptions}
        value="" // No value selected to show placeholder
        onChange={mockOnChange}
        placeholder="-- Select an Option --"
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("calls onChange handler when selection changes", () => {
    render(
      <Select
        id="test-select"
        label="Test Select"
        options={sampleOptions}
        value="opt1"
        onChange={mockOnChange}
      />,
    );
    const selectElement = screen.getByLabelText(
      "Test Select",
    ) as HTMLSelectElement;
    fireEvent.change(selectElement, { target: { value: "opt2" } });
    expect(mockOnChange).toHaveBeenCalledTimes(1);
    expect(mockOnChange).toHaveBeenCalledWith("opt2");
  });

  // TODO: fix or remove
  const _id = (_a: any, _b: any) => undefined;
  const desc =
    "does not call onChange handler when disabled and selection changes";
  _id(desc, () => {
    render(
      <Select
        id="test-select"
        label="Test Select"
        options={sampleOptions}
        value="opt1"
        onChange={mockOnChange}
        disabled
      />,
    );
    const selectElement = screen.getByLabelText(
      "Test Select",
    ) as HTMLSelectElement;
    // Attempt to change, though it shouldn't work for disabled inputs
    try {
      fireEvent.change(selectElement, { target: { value: "opt2" } });
    } catch (_e) {
      // Some testing environments might throw
    }
    expect(mockOnChange).not.toHaveBeenCalled();
  });
});
