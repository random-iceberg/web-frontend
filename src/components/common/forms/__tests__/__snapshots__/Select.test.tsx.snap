// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Component renders correctly when disabled 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <label
      class="block text-sm font-medium text-gray-700 mb-2"
      for="test-select"
    >
      Test Select
    </label>
    <select
      class="w-full rounded border-gray-300 shadow-sm text-sm px-4 py-2.5
          focus:ring-1 focus:ring-blue-500 focus:border-blue-500
          disabled:bg-gray-50 disabled:text-gray-500 disabled:cursor-not-allowed
          border"
      disabled=""
      id="test-select"
    >
      <option
        value="opt1"
      >
        Option 1
      </option>
      <option
        value="opt2"
      >
        Option 2
      </option>
      <option
        value="opt3"
      >
        Option 3
      </option>
    </select>
  </div>
</DocumentFragment>
`;

exports[`Select Component renders correctly when required 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <label
      class="block text-sm font-medium text-gray-700 mb-2"
      for="test-select"
    >
      Test Select
      <span
        class="text-red-500 ml-1"
      >
        *
      </span>
    </label>
    <select
      class="w-full rounded border-gray-300 shadow-sm text-sm px-4 py-2.5
          focus:ring-1 focus:ring-blue-500 focus:border-blue-500
          disabled:bg-gray-50 disabled:text-gray-500 disabled:cursor-not-allowed
          border"
      id="test-select"
      required=""
    >
      <option
        value="opt1"
      >
        Option 1
      </option>
      <option
        value="opt2"
      >
        Option 2
      </option>
      <option
        value="opt3"
      >
        Option 3
      </option>
    </select>
  </div>
</DocumentFragment>
`;

exports[`Select Component renders correctly with a different selected value 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <label
      class="block text-sm font-medium text-gray-700 mb-2"
      for="test-select"
    >
      Test Select
    </label>
    <select
      class="w-full rounded border-gray-300 shadow-sm text-sm px-4 py-2.5
          focus:ring-1 focus:ring-blue-500 focus:border-blue-500
          disabled:bg-gray-50 disabled:text-gray-500 disabled:cursor-not-allowed
          border"
      id="test-select"
    >
      <option
        value="opt1"
      >
        Option 1
      </option>
      <option
        value="opt2"
      >
        Option 2
      </option>
      <option
        value="opt3"
      >
        Option 3
      </option>
    </select>
  </div>
</DocumentFragment>
`;

exports[`Select Component renders correctly with basic props 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <label
      class="block text-sm font-medium text-gray-700 mb-2"
      for="test-select"
    >
      Test Select
    </label>
    <select
      class="w-full rounded border-gray-300 shadow-sm text-sm px-4 py-2.5
          focus:ring-1 focus:ring-blue-500 focus:border-blue-500
          disabled:bg-gray-50 disabled:text-gray-500 disabled:cursor-not-allowed
          border"
      id="test-select"
    >
      <option
        value="opt1"
      >
        Option 1
      </option>
      <option
        value="opt2"
      >
        Option 2
      </option>
      <option
        value="opt3"
      >
        Option 3
      </option>
    </select>
  </div>
</DocumentFragment>
`;

exports[`Select Component renders with a placeholder 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <label
      class="block text-sm font-medium text-gray-700 mb-2"
      for="test-select"
    >
      Test Select
    </label>
    <select
      class="w-full rounded border-gray-300 shadow-sm text-sm px-4 py-2.5
          focus:ring-1 focus:ring-blue-500 focus:border-blue-500
          disabled:bg-gray-50 disabled:text-gray-500 disabled:cursor-not-allowed
          border"
      id="test-select"
    >
      <option
        disabled=""
        value=""
      >
        -- Select an Option --
      </option>
      <option
        value="opt1"
      >
        Option 1
      </option>
      <option
        value="opt2"
      >
        Option 2
      </option>
      <option
        value="opt3"
      >
        Option 3
      </option>
    </select>
  </div>
</DocumentFragment>
`;
